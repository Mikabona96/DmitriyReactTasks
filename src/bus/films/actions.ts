// API
import { api } from '../../api';

//Types
import { types } from './types';

export const filmsActions = Object.freeze({
    //Sync
    startFetching: ()=>{
        return {
            type: types.FILM_START_FETCHING,
        };
    },
    stopFetching: ()=>{
        return {
            type: types.FILM_STOP_FETCHING,
        };
    },
    fill: (payload: any)=>{
        return {
            type: types.FILM_FILL,
            payload,
        };
    },
    setCurrentFilm: (payload: any)=>{
        return {
            type: types.FILM_SET_CURRENT_FILM,
            payload,
        };
    },
    setFetchingError: (error: any)=>{
        return {
            type:    types.FILM_SET_FETCHING_ERROR,
            error:   true,
            payload: error,
        };
    },

    fetchAsync: () => async(dispatch: any)=>{
        dispatch(filmsActions.startFetching());

        const response = await api.films.fetch();

        if (response.status === 200) {
            const { results } = await response.json();

            dispatch(filmsActions.fill(results));
        } else {
            const error = {
                status: response.status,
            };

            dispatch(filmsActions.setFetchingError(error));
        }
        dispatch(filmsActions.stopFetching());
    },
});
