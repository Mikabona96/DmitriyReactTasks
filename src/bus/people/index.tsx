/* eslint-disable react/no-unused-prop-types */
import React, { FC } from 'react';
import { usePeopleFetch } from './hooks/usePeopleFetch';
import { Outlet, useNavigate } from 'react-router-dom';
import { PeopleSection, PeoplePopup, ListJSX } from './styles';
import { useDispatch } from 'react-redux';
import { peopleActions } from './actions';

export const People: FC = (): any=>{
    const peopleFetch = usePeopleFetch();
    const navigate = useNavigate();
    const dispatch = useDispatch();
    if (peopleFetch) {
        const { isFetching, data, error } = peopleFetch;
        if (error && error.status === 404) {
            return <p>Not Found !</p>;
        }

        if (error && error.status !== 404) {
            return <p>Something went wrong</p>;
        }

        const spinnerJSX = isFetching && (
            <p>Loading data from api</p>
        );

        const listJSX = !isFetching && data && data.length && data.map((person: any, index: number)=>{
            const idRegExp = /\/([0-9]*)\/$/;
            const id =  person.url.match(idRegExp)[ 1 ];

            return (
                <ListJSX
                    key = { String(index) }
                    onClick = { () => {
                        dispatch(peopleActions.setCurrentPerson(person));
                        navigate(`/lesson7/people/${id}`);
                    } }>{person.name}
                </ListJSX>
            );
        });

        return (
            <PeopleSection>
                <div>
                    {spinnerJSX}
                    {listJSX}
                </div>
                <PeoplePopup>
                    <Outlet />
                </PeoplePopup>
            </PeopleSection>
        );
    }

    return <div>Ooops</div>;
};
