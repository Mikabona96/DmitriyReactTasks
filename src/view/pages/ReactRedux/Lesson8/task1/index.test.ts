import { peopleActions } from '../../../../../bus/people/actions';


test('startFetching ahould return Object {type: PEOPLE_START_FETCHING}', () => {
    expect(peopleActions.startFetching()).toMatchSnapshot();
});
test('stopFetching ahould return Object {type: PEOPLE_STOP_FETCHING}', () => {
    expect(peopleActions.stopFetching()).toMatchSnapshot();
});
test('fill ahould return Object {type: PEOPLE_FILL, payload}', () => {
    expect(peopleActions.fill('er')).toMatchSnapshot();
});
test('setCurrentPerson ahould return {type: PEOPLE_START_FETCHING, payload}', () => {
    expect(peopleActions.setCurrentPerson('er')).toMatchSnapshot();
});
test('setFetchingError ahould return {type: PEOPLE_START_FETCHING}', () => {
    expect(peopleActions.setFetchingError(404)).toMatchSnapshot();
});
