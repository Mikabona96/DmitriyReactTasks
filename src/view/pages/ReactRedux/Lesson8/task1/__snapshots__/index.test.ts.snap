// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fill ahould return Object {type: PEOPLE_FILL, payload} 1`] = `
Object {
  "payload": "er",
  "type": "PEOPLE_FILL",
}
`;

exports[`setCurrentPerson ahould return {type: PEOPLE_START_FETCHING, payload} 1`] = `
Object {
  "payload": "er",
  "type": "PEOPLE_SET_CURRENT_PERSON",
}
`;

exports[`setFetchingError ahould return {type: PEOPLE_START_FETCHING} 1`] = `
Object {
  "error": true,
  "payload": 404,
  "type": "PEOPLE_SET_FETCHING_ERROR",
}
`;

exports[`startFetching ahould return Object {type: PEOPLE_START_FETCHING} 1`] = `
Object {
  "type": "PEOPLE_START_FETCHING",
}
`;

exports[`stopFetching ahould return Object {type: PEOPLE_STOP_FETCHING} 1`] = `
Object {
  "type": "PEOPLE_STOP_FETCHING",
}
`;
